rails/18_project_auth.md

```diff
@@ -0,0 +1,99 @@
+
+6. Создайте несколько пользователей чтобы заполнить Ваше приложение. Мы не будем писать регистрационную форму, так что нужно создавать новых пользователей через командную строку. Ваш `#before_create` сейчас должен правильно присваивать новым пользователям специальный токен.
+7. Сейчас заполните в действии `#create` в Вашем SessionsController чтобы на самом деле создать сессию пользователя. Сначала нужно найти пользователя основываясь на адресе email, а затем сравнить хэш пароля , который он отправил в параметрах, с хэшированным паролем хранящимся в базе данных (используя `#authenticate`).  Смотрите [Главу 8](http://ruby.railstutorial.org/chapters/sign-in-sign-out#sec-reviewing_form_submission) с вопросами, но попытайтесь не копировать сразу дословно - Вы делаете это чтобы научиться.
+8. Как только Вы подтвердите, что пользователь ввел правильный пароль, впустите его.
+9. Создайте новый метод в ApplicationController который выполняет вход для Вас. Присвойте пользователю новый токен (так чтобы он не был чужим или устаревшим). Храните токен в браузере пользователя используя cookie, так всякий раз когда он посещяет новую страницу, мы можем проверять выполнен ли вход. Используйте `cookies.permanent` "hash" чтобы сделать это.
+10. Создайте 2 других полезных метода в ApplicationController -- один чтобы возвращать текущего пользователя (`#current_user`) и другой чтобы устанавливать его (`#current_user=(user)`). В возвращении текущего пользователя должен использоваться оператор `||=` -- если текущий пользователь уже установлен, Вы должны вернуть этого пользователя, иначе, Вам нужно вытянуть токен из cookie и искать его в базе данных чтобы вытащить соответствующего пользователя.  Если пользователь не найден, то верните `nil`.
+11. Устанавливайте текущего пользователя всегда когда пользователь осуществляет вход.
+12. Напишите функцию выхода в действии `SessionsController#delete`  которая будет убирать текущего пользователя и удалять токен с cookie. Для этого лучше всего вызвать метод (например `#sign_out`) в ApplicationController вместо того чтобы просто писать весь функционал в SessionsController.
+13. Создайте ссылку для выхода, которая будет вызывать метод `#delete` в контроллере сессии.  Понадобиться обманывать метод DELETE HTTP, но это легко делается путем передачи в `#link_to` опции `:method => :delete`.
+
+#### Атентификация и посты
+
+Давайте напишем эти секреты! Мы должны быть уверены, что видеть автора каждого поста могут только те пользователи, которые выполнили вход. Мы не собираемся волноваться о редактировании и удалении постов.
+
+1. Создайте Post model и Posts controller и соответствующий ресурс в Routes файле, который позволяет использовать методы `[:new, :create, :index]`.
+2. Поверх  Posts Controller, используй a `#before_filter` чтобы ограничить доступ к методам `#new` и`#create` только для участников, которые выполнили вход. Создайте обязательные вспомогательные методы в ApplicationController.
```

Comment:
"Atop your Posts Controller, use a #before_filter to restrict access to the #new and #create methods to only users who are signed in."